SPEC

```diff
@@ -119,17 +119,17 @@ Servers that choose to support WebSockets and SSE connections should follow thes
                          :websocket for a WebSocket upgrade, or :sse for an SSE upgrade.
 <tt>rack.upgrade</tt>:: Is used to pass a handler instance or class back to the server for WebSocket and SSE connections.
                      The handler MAY implement any or all of the following callbacks:
-                      on_open()                        # called when the connection is complete
-                      on_message(message)              # called when a WebSocket message is received by server. <tt>message</tt> will be UTF-8 encoded for text messages and Binary encoded for binary messages.
-                      on_shutdown()                    # may be called before a connection is closed due to server shutdown.
-                      on_close()                       # called when the connection is closed
-                      on_drained                       # may be called when the number of pending writes drops to zero.
-                     The object will be extended by the server to include:
-                      write(message)  # writes to the WebSocket or SSE connection
-                      close()         # forces a close of the WebSocket or SSE connection
-                      open?()         # returns true if the connection is open, false otherwise
-                      pending()       # returns the number of pending writes or -1 if the connection is closed
-                     This option should only be set it the environment options has a non-nil value for the <tt>rack.upgrade?</tt> option,
+                      on_open(client)      # called when the connection is complete with a client object.
+                      on_message(message)  # called when a WebSocket message is received by server. <tt>message</tt> will be UTF-8 encoded for text messages and Binary encoded for binary messages.
+                      on_shutdown          # may be called before a connection is closed due to server shutdown.
+                      on_close             # called when the connection is closed
```

Comment:
I think it's only fair that the API informs the receiver about the identity of the closed client (i.e., for `collection.delete client` type of situations)?

so `on_close(client)` as well.

Reply:
Oh, sorry, I misunderstood... I think this update assumes the client object is stored by the receiver.

Reply:
The call to `on_close` indicates the connection has been closed so the client is no longer viable or at least any method called on it would raise an exception as not being connected so it doesn't seem useful to include it as an argument to the close callback.

Reply:
Did miss adding the client to the `on_message`. Fixed that.

Reply:
On second though. I hadn't fully bend my mind around what the change would mean. With a stateless approach as @tenderlove suggested a single upgrade handler could be used. It might then want to store an array of clients thereby avoiding creating a two new objects for every connection. Maybe thats what you were thinking of before my hasty response.

I'll update the spec with all calls taking the client.